CRM/Contribute/Form/Task/Invoice.php

```diff
@@ -543,7 +543,12 @@ public static function printPDF($contribIDs, &$params, $contactIds) {
         self::addActivities($subject, $contribution->contact_id, $fileName, $params);
       }
 
-      CRM_Core_DAO::setFieldValue('CRM_Contribute_DAO_Contribution', $contribution->id, 'invoice_id', $invoiceId);
+      // CRM-20387 save my human-readable invoice number as invoice_id in the contribution table,
+      // but only if the current value is one of: set by this invoicing code, or pay-later, or empty.
+      $set_by_invoicing_code = (0 === strpos($contribution->invoice_id, CRM_Utils_Array::value('invoice_prefix', $prefixValue)));
```

Comment:
Actually that's a good thought experiment - what if a site did change their prefix and then re-issued an invoice? In the current code, the invoice number would be changed on the printed/emailed pdf as well as in the database, and no record of the previous invoice or it's (different) name would exist. In fact, there would be a misleading activity from the initial invoice issuance that would only link to a copy of the new invoice pdf. So this patch actually addresses the problem you are describing, which is a problem with the current code, it's not creating the problem.

The key question this issue raises is really: what is/was the point of rewriting the invoice_id db field? That field is not used in generating the human-readable invoice, and as far as I can tell there is no benefit, and has several risks (both the reconciliation problem that spurred this patch, as well as likely violating some UK invoicing rule). Putting the human-readable invoice string into this field doesn't give you any reliable searchability benefit (e.g. the example you just provided), and in any case, you can use the contribution id to find the matching contribution if you are holding an invoice in front of you. So I'm still hoping someone will explain why the code does this in case there is some functionality hidden in there. But at this point, my best guess is that it was an oversight based on an incomplete understanding of how invoice_id is used in core and payment processors.

And lastly - yes, there are a two other improvements to supplement this: support for partial payments is now in, and versioning/tracking of invoices is proposed.